wemake_python_styleguide/visitors/ast/function_empty_lines.py

```diff
@@ -8,53 +9,124 @@
+        return is_dedent_token and function_tokens
+
+
+@final
+class _FileTokens(object):
+
+    def __init__(
+        self,
+        file_functions: _FileFunctions,
+        exps_for_one_empty_line: int,
+    ) -> None:
+        self._file_functions = file_functions
+        self._exps_for_one_empty_line = exps_for_one_empty_line
+
+    def analyze(self) -> List[best_practices.WrongEmptyLinesCountViolation]:
+        violations = []
+        for function in self._file_functions.as_list():
+            splitted_function_body = function.body().strip().split('\n')
+            empty_lines_count = len([
+                line for line in splitted_function_body if line == ''
+            ])
+            if not empty_lines_count:
+                return []
+            available_empty_lines = self._available_empty_lines(
+                len(splitted_function_body), empty_lines_count,
+            )
+            if empty_lines_count > available_empty_lines:
+                violations.append(
+                    best_practices.WrongEmptyLinesCountViolation(
+                        function.name_token(),
+                        text=str(empty_lines_count),
+                        baseline=available_empty_lines,
+                    ),
                 )
-                if self._empty_lines_count > available_empty_lines:
-                    self.add_violation(
-                        best_practices.WrongEmptyLinesCountViolation(
-                            token,
-                            text=str(self._empty_lines_count),
-                            baseline=available_empty_lines,
-                        ),
-                    )
-                self._function_start_line = 0
-
-    def _try_mark_function_start(self, token: tokenize.TokenInfo):
-        if token.string == 'def':
-            self._empty_lines_count = 0
-            self._function_start_line = token.start[0]
+        return violations
 
     def _available_empty_lines(
         self,
         function_len: int,
         empty_lines: int,
     ) -> int:
-        option = self.options.exps_for_one_empty_line
+        option = self._exps_for_one_empty_line
         if option == 0:
             return 0
         lines_with_expressions = function_len - empty_lines
         return math.floor(lines_with_expressions / option)
+
+
+@final
+class WrongEmptyLinesCountVisitor(base.BaseTokenVisitor):
+    """Restricts empty lines in function or method body."""
+
+    def __init__(self, *args, **kwargs) -> None:
+        """Initializes a counter."""
+        super().__init__(*args, **kwargs)
+        self._file_tokens: List[tokenize.TokenInfo] = []
+
+    def visit(self, token: tokenize.TokenInfo) -> None:
+        """Find empty lines count."""
+        self._file_tokens.append(token)
+        if token.type != tokenize.ENDMARKER:
+            return
+        violations = _FileTokens(
+            _FileFunctions(
+                tuple(self._file_tokens),
```

Comment:
Why do we need this cast? Why not `list`?

Reply:
I used tuple, because it provide immutability between method calls. If you think the list will be enough, `_FileFunctions` constructor will accept the list

Reply:
Please

Reply:
Okey, changed